### ISSUE TITLE
[Phase 5] Implement MeanReversionStrategy

### ISSUE THREAD

**Issue by kimrejstrom** [kimrejstrom/alpacalyzer-algo-trader#25]
## Overview

Implement a `MeanReversionStrategy` that identifies overbought/oversold conditions and trades the reversion to the mean.

## Background

Mean reversion assumes that prices tend to return to their average over time. This strategy enters when price deviates significantly from the mean, expecting a reversal.

## Strategy Logic

Entry conditions (Long):
1. RSI below oversold threshold (< 30)
2. Price below lower Bollinger Band
3. Volume spike indicating capitulation
4. Not in strong downtrend

Entry conditions (Short):
1. RSI above overbought threshold (> 70)
2. Price above upper Bollinger Band
3. Volume spike indicating exhaustion
4. Not in strong uptrend

Exit conditions:
1. Price returns to mean (SMA)
2. RSI normalizes
3. Stop loss (extended deviation)
4. Time-based exit

## Tasks

- [ ] Create `src/alpacalyzer/strategies/mean_reversion.py`
- [ ] Implement `MeanReversionStrategy` class
- [ ] Add RSI and Bollinger Band calculations
- [ ] Create mean reversion detection
- [ ] Write unit tests

## Implementation

```python
# src/alpacalyzer/strategies/mean_reversion.py

from dataclasses import dataclass
from datetime import datetime, UTC, timedelta
from typing import Literal

import pandas as pd
import numpy as np

from alpacalyzer.strategies.base import Strategy, StrategyConfig, EntrySignal, ExitSignal


@dataclass
class MeanReversionConfig(StrategyConfig):
    """Configuration for mean reversion strategy."""
    name: str = "mean_reversion"
    
    # RSI parameters
    rsi_period: int = 14
    rsi_oversold: float = 30.0
    rsi_overbought: float = 70.0
    rsi_exit_threshold: float = 50.0  # Exit when RSI crosses this
    
    # Bollinger Bands
    bb_period: int = 20
    bb_std: float = 2.0
    
    # Mean parameters
    mean_period: int = 20
    deviation_threshold: float = 2.0  # Std devs from mean
    
    # Risk management
    risk_pct_per_trade: float = 0.015
    max_hold_hours: int = 48  # Force exit after this time
    stop_loss_std: float = 3.0  # Stop at this many std devs
    
    # Filters
    min_volume_ratio: float = 1.2
    trend_filter_period: int = 50


class MeanReversionStrategy(Strategy):
    """
    Mean reversion trading strategy.
    
    Identifies extreme deviations from the mean and trades
    the expected reversion.
    """
    
    def __init__(self, config: MeanReversionConfig | None = None):
        self.config = config or MeanReversionConfig()
    
    @property
    def name(self) -> str:
        return self.config.name
    
    def should_enter(
        self,
        ticker: str,
        price: float,
        df: pd.DataFrame,
    ) -> tuple[bool, EntrySignal | None]:
        """Check if mean reversion conditions are met."""
        
        if len(df) < max(self.config.rsi_period, self.config.bb_period, self.config.trend_filter_period) + 10:
            return False, None
        
        # Calculate indicators
        rsi = self._calculate_rsi(df)
        bb_upper, bb_middle, bb_lower = self._calculate_bollinger_bands(df)
        z_score = self._calculate_z_score(df)
        
        latest = df.iloc[-1]
        current_rsi = rsi.iloc[-1]
        
        # Check volume
        avg_volume = df["volume"].tail(50).mean()
        current_volume = latest["volume"]
        volume_ratio = current_volume / avg_volume if avg_volume > 0 else 0
        
        if volume_ratio < self.config.min_volume_ratio:
            return False, None
        
        # Trend filter - avoid trading against strong trends
        sma_long = df["close"].tail(self.config.trend_filter_period).mean()
        sma_short = df["close"].tail(20).mean()
        trend_strength = (sma_short - sma_long) / sma_long
        
        # Oversold - Long opportunity
        if (
            current_rsi < self.config.rsi_oversold
            and price < bb_lower.iloc[-1]
            and z_score < -self.config.deviation_threshold
            and trend_strength > -0.10  # Not in strong downtrend
        ):
            mean = bb_middle.iloc[-1]
            std = (bb_upper.iloc[-1] - bb_middle.iloc[-1]) / self.config.bb_std
            stop_loss = price - (std * self.config.stop_loss_std)
            target = mean  # Target is the mean
            
            return True, EntrySignal(
                ticker=ticker,
                side="long",
                price=price,
                stop_loss=stop_loss,
                target=target,
                confidence=self._calculate_confidence(current_rsi, z_score, "oversold"),
                reason=f"Oversold: RSI={current_rsi:.1f}, Z-score={z_score:.2f}, below BB",
            )
        
        # Overbought - Short opportunity
        if (
            current_rsi > self.config.rsi_overbought
            and price > bb_upper.iloc[-1]
            and z_score > self.config.deviation_threshold
            and trend_strength < 0.10  # Not in strong uptrend
        ):
            mean = bb_middle.iloc[-1]
            std = (bb_upper.iloc[-1] - bb_middle.iloc[-1]) / self.config.bb_std
            stop_loss = price + (std * self.config.stop_loss_std)
            target = mean
            
            return True, EntrySignal(
                ticker=ticker,
                side="short",
                price=price,
                stop_loss=stop_loss,
                target=target,
                confidence=self._calculate_confidence(current_rsi, z_score, "overbought"),
                reason=f"Overbought: RSI={current_rsi:.1f}, Z-score={z_score:.2f}, above BB",
            )
        
        return False, None
    
    def should_exit(
        self,
        ticker: str,
        entry_price: float,
        current_price: float,
        position_side: str,
        df: pd.DataFrame,
        **kwargs,
    ) -> tuple[bool, ExitSignal | None]:
        """Check if position should be exited."""
        
        stop_loss = kwargs.get("stop_loss", 0)
        target = kwargs.get("target", 0)
        entry_time = kwargs.get("entry_time")
        
        rsi = self._calculate_rsi(df)
        current_rsi = rsi.iloc[-1]
        
        # Stop loss
        if position_side == "long" and current_price <= stop_loss:
            return True, ExitSignal(
                ticker=ticker,
                side=position_side,
                price=current_price,
                reason="stop_loss",
                urgency="immediate",
            )
        
        if position_side == "short" and current_price >= stop_loss:
            return True, ExitSignal(
                ticker=ticker,
                side=position_side,
                price=current_price,
                reason="stop_loss",
                urgency="immediate",
            )
        
        # Mean reversion achieved (target)
        if position_side == "long" and current_price >= target:
            return True, ExitSignal(
                ticker=ticker,
                side=position_side,
                price=current_price,
                reason="target_reached",
                urgency="normal",
            )
        
        if position_side == "short" and current_price <= target:
            return True, ExitSignal(
                ticker=ticker,
                side=position_side,
                price=current_price,
                reason="target_reached",
                urgency="normal",
            )
        
        # RSI normalization exit
        rsi_normalized = self.config.rsi_oversold < current_rsi < self.config.rsi_overbought
        crossed_middle = abs(current_rsi - self.config.rsi_exit_threshold) < 5
        
        if rsi_normalized and crossed_middle:
            return True, ExitSignal(
                ticker=ticker,
                side=position_side,
                price=current_price,
                reason="rsi_normalized",
                urgency="normal",
            )
        
        # Time-based exit
        if entry_time:
            hold_duration = datetime.now(UTC) - entry_time
            if hold_duration > timedelta(hours=self.config.max_hold_hours):
                return True, ExitSignal(
                    ticker=ticker,
                    side=position_side,
                    price=current_price,
                    reason="max_hold_time",
                    urgency="urgent",
                )
        
        return False, None
    
    def _calculate_rsi(self, df: pd.DataFrame) -> pd.Series:
        """Calculate Relative Strength Index."""
        delta = df["close"].diff()
        gain = delta.where(delta > 0, 0)
        loss = -delta.where(delta < 0, 0)
        
        avg_gain = gain.rolling(window=self.config.rsi_period).mean()
        avg_loss = loss.rolling(window=self.config.rsi_period).mean()
        
        rs = avg_gain / avg_loss
        rsi = 100 - (100 / (1 + rs))
        
        return rsi
    
    def _calculate_bollinger_bands(self, df: pd.DataFrame) -> tuple[pd.Series, pd.Series, pd.Series]:
        """Calculate Bollinger Bands."""
        middle = df["close"].rolling(window=self.config.bb_period).mean()
        std = df["close"].rolling(window=self.config.bb_period).std()
        
        upper = middle + (std * self.config.bb_std)
        lower = middle - (std * self.config.bb_std)
        
        return upper, middle, lower
    
    def _calculate_z_score(self, df: pd.DataFrame) -> float:
        """Calculate Z-score of current price from mean."""
        mean = df["close"].tail(self.config.mean_period).mean()
        std = df["close"].tail(self.config.mean_period).std()
        
        if std == 0:
            return 0.0
        
        return (df.iloc[-1]["close"] - mean) / std
    
    def _calculate_confidence(self, rsi: float, z_score: float, condition: str) -> float:
        """Calculate confidence score."""
        confidence = 50.0
        
        if condition == "oversold":
            # Lower RSI = higher confidence
            if rsi < 20:
                confidence += 20
            elif rsi < 25:
                confidence += 15
            elif rsi < 30:
                confidence += 10
            
            # More extreme Z-score = higher confidence
            if z_score < -3:
                confidence += 15
            elif z_score < -2.5:
                confidence += 10
            
        else:  # overbought
            if rsi > 80:
                confidence += 20
            elif rsi > 75:
                confidence += 15
            elif rsi > 70:
                confidence += 10
            
            if z_score > 3:
                confidence += 15
            elif z_score > 2.5:
                confidence += 10
        
        return min(confidence, 95.0)
```

## Acceptance Criteria

- [ ] `MeanReversionStrategy` class implements `Strategy` protocol
- [ ] `MeanReversionConfig` with all parameters
- [ ] RSI calculation and thresholds
- [ ] Bollinger Band integration
- [ ] Z-score calculation
- [ ] Time-based exit
- [ ] Trend filter to avoid fighting trends
- [ ] Unit tests for all conditions
- [ ] Backtest performance validation

## Dependencies

- #5 Strategy Protocol
- #6 Strategy Registry
- #25 Backtesting Framework

## Estimated Effort

3-4 hours
